### PR TITLE
Implement assignT and unassignT

### LINKED ISSUE #147: Implement AssignT and UnassignT

### PR DESCRIPTION
**Pull request by rayyan35p**
Closes #147

Commands implemented:

`assignT `
`unassignT `

Example usage:

`assignT pr/1 t/2 em/3`
`unassignT pr/1 t/2 em/3`

Notes:

- Take note that the project index refers to the index of projects as per the Projects list. However, the indexes of the task and employee must refer to the indexes of the corresponding elements in that particular project.

- The GUI has been updated to display indexes of members in the project to accommodate this:

image

- In summary, the task index and employee index is to be retrieved from their indexes in the project card.

- The commands (as of now) can only assign/unassign one employee from a task, i.e. `Task` can only store one employee.

- After merging this PR, users can assign and unassign tasks to/from employees within a project.

### CODE REVIEW

File: src/main/java/seedu/address/model/task/Task.java [hunk 1]
```diff
@@ -47,23 +51,64 @@ public Task(String taskName, LocalDateTime deadline, Boolean isDone) {
...
+        this.employee = Collections.singletonList(employee);
+    }
+
+    /**
+     * Constructs a Task with a pre-determined Done status and assigned employee
+     * - for loading of tasks from taskhub.json.
+     * @param taskName The name of the task.
+     * @param employee The employee that was assigned to the task.
+     */
+    public Task(String taskName, LocalDateTime deadline, Boolean isDone, Employee employee) {
```
Comment (aslam341) [inline src/main/java/seedu/address/model/task/Task.java:79]: Perhaps `boolean` primitive type is better than `Boolean` reference type, since `Boolean` can hold null values and this will cause a `NullPointerException` during the auto-boxing. It is better to prevent the "error" during compile-time than catch it during runtime. There is another constructor in this class with the same potential problem.

File: src/main/java/seedu/address/model/task/Task.java [hunk 2]
```diff
@@ -47,23 +51,64 @@ public Task(String taskName, LocalDateTime deadline, Boolean isDone) {
...
+     * - for loading of tasks from taskhub.json
+     * @param taskName The name of the task.
+     * @param employee The employee that was assigned to the task.
      */
-    public static boolean isValidTask(String testString) {
-        return testString.matches(VALIDATION_REGEX);
+    public Task(String taskName, LocalDateTime deadline, Employee employee) {
+        requireNonNull(taskName);
+        requireNonNull(deadline);
+        requireNonNull(employee);
```
Comment (aslam341) [inline src/main/java/seedu/address/model/task/Task.java:66]: There is a 'requireAllNonNull` method, which will shorten the lines here from 3 to 1. Likewise for the other constructor methods in this class. I believe you're aware of this though, since you used the method in another class, so is there a reason for separating it here?

File: src/main/java/seedu/address/storage/JsonAdaptedTask.java [hunk 3]
```diff
@@ -28,10 +31,14 @@ public class JsonAdaptedTask {
     @JsonCreator
     public JsonAdaptedTask(@JsonProperty("name") String name,
                            @JsonProperty("deadline") LocalDateTime deadline,
-                           @JsonProperty("isDone") Boolean isDone) {
+                           @JsonProperty("isDone") Boolean isDone,
```
Comment (aslam341) [inline src/main/java/seedu/address/storage/JsonAdaptedTask.java:34]: Similar to the previous comment, better to use `boolean` instead of `Boolean`.

File: src/main/java/seedu/address/logic/commands/AssignTaskCommand.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+ * Assigns a task in a project to employee(s) in that project.
+ */
+public class AssignTaskCommand extends Command {
+    public static final String COMMAND_WORD = "assignT";
+    public static final String MESSAGE_SUCCESS = "Assigned task: %1$s to employee, %2$s";
+    public static final String MESSAGE_USAGE =
+            COMMAND_WORD + ": Assigns a task in a project to an employee in the project.\n"
+            + "Parameters: "
+            + PREFIX_PROJECT + "PROJECT_INDEX "
+            + PREFIX_TAG + "TASK_INDEX "
```
Comment (aslam341) [inline src/main/java/seedu/address/logic/commands/AssignTaskCommand.java:32]: I think it would be better to create and use PREFIX_TASK instead of PREFIX_TAG. It does not matter that both of them share the same prefix "/t" currently, and if we decide to change one of the prefix, the other will still remain the same.

**Review by aslam341 — COMMENTED** [review]
Just leaving some comments for now. I have yet to test the code, and these comments are based on a quick look through of the code. I may give more comments later on when I have the time.

File: src/main/java/seedu/address/logic/commands/AssignTaskCommand.java (hunk 4 above)
Comment (rayyan35p) [reply src/main/java/seedu/address/logic/commands/AssignTaskCommand.java:32]: Ah yes, I was already changing them to task but I must have missed these, thanks!

File: src/main/java/seedu/address/model/task/Task.java (hunk 1 above)
Comment (rayyan35p) [reply src/main/java/seedu/address/model/task/Task.java:79]: Actually I was just using the existing implementation of the constructor which should have either been done by @timleow who made the class, or @Chandan8186 who handled the marking of tasks. Perhaps they would be better suited to comment on this instead of me, although I do agree with you.

File: src/main/java/seedu/address/model/task/Task.java (hunk 2 above)
Comment (rayyan35p) [reply src/main/java/seedu/address/model/task/Task.java:66]: No reason, I must have missed this. Thanks!

**Comment by rayyan35p** [comment]
@aslam341 @Chandan8186 Do let me know your thoughts on how I implemented the assigned employee attribute in task. I wasn't able to use just an `Employee` class because if there was no assigned employee, it would have been null hence why I opted to use a List instead, similar to `Tag`. Do you guys think it's necessary to make a wrapper class for this since the behavior differs from a normal list? (Only one item in the list for example)

**Comment by antonTan96** [comment]
Would using an Optional<Employee> be better instead of using a list? Since a task may or may not have an employee. However, if we plan to assign multiple employees to a task then using a list would be better.

**Comment by rayyan35p** [comment]
> Would using an Optional be better instead of using a list? Since a task may or may not have an employee. However, if we plan to assign multiple employees to a task then using a list would be better.

Honestly, I forgot about Optional so I didn't even consider it so thanks for the suggestion but yeah you do raise a good point as well about how it would be better to keep it as List if we do intend on expanding tasks to take more employees.

**Comment by Chandan8186** [comment]
Yes, I agree with @antonTan96 on using `Optional`/`List`, depending on the multiplicity of the Employee association (I had mentioned something similar earlier regarding `ProjectDeadline`). If you are only allowing one employee to be added to a task, it might be a bit safer to use `Optional`, but at the same time, you can have proper guards against multiple employees when using `List`. I will add more comments if necessary during my PR review.

File: src/main/java/seedu/address/logic/commands/UnassignTaskCommand.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        editedTaskList.setTask(taskIndex, editedTask);
+        Project editedProject = new Project(projectToEdit.getNameString(),
+                projectToEdit.getEmployees(),
+                editedTaskList,
+                projectToEdit.getProjectPriority(),
+                projectToEdit.getDeadline(),
+                projectToEdit.getCompletionStatus());
+        model.setProject(projectToEdit, editedProject);
+        model.updateFilteredProjectList(Model.PREDICATE_SHOW_ALL_PROJECTS);
+        Employee assignedEmployee = taskToEdit.getEmployee().get(0);
```
Comment (Chandan8186) [inline src/main/java/seedu/address/logic/commands/UnassignTaskCommand.java:77]: if there is no `isEmpty` check for the employee list, there will be an uncaught `IndexOutOfBoundsException`. I will provide more details in my overall PR review summary comment.

**Review by Chandan8186 — CHANGES_REQUESTED** [review]
LGTM, other than the minor change suggested for the `assignT` and `unassignT` messages. However, I have noticed a bug: when I unassign from a task that doesn't have an existing employee, the input field does not clear and there is no reply message displayed. On the terminal, there seems to be an uncaught `IndexOutOfBoundsException`, which suggests that there seems to be a missing isEmpty guard for the task's employeeList.

image

File: src/main/java/seedu/address/logic/commands/AssignTaskCommand.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+import seedu.address.model.task.TaskList;
+
+/**
+ * Assigns a task in a project to employee(s) in that project.
+ */
+public class AssignTaskCommand extends Command {
+    public static final String COMMAND_WORD = "assignT";
+    public static final String MESSAGE_SUCCESS = "Assigned task: %1$s to employee, %2$s";
+    public static final String MESSAGE_USAGE =
+            COMMAND_WORD + ": Assigns a task in a project to an employee in the project.\n"
```
[the remark this answers is not in the document]
Comment (rayyan35p) [reply src/main/java/seedu/address/logic/commands/AssignTaskCommand.java:29]: Yep, that sounds clearer.

File: src/main/java/seedu/address/model/task/TaskList.java [hunk 7]
```diff
@@ -98,12 +98,28 @@ public void setTask(Task target, Task editedTask) {
         }
 
         if (!target.equals(editedTask) && contains(editedTask)) {
-            throw new DuplicateProjectException();
+            throw new DuplicateTaskException();
         }
```
[the remark this answers is not in the document]
Comment (rayyan35p) [reply src/main/java/seedu/address/model/task/TaskList.java:102]: It's alright, I'll fix it in this PR!

**Review by aslam341 — APPROVED** [review]
LGTM.

**Review by Chandan8186 — APPROVED** [review]
LGTM!
